Modify PyTorch's integration of NNPACK to use a unified underlying thread pool implementation.

AshkanAliabadi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #27402 Use NNPACK for strided convolutions.
* **#27341 Modify PyTorch's integration of NNPACK to use a unified underlying thread pool implementation.**

Multi-threaded:

```
Pixel 2:
Before:    362.716
PR-27402:  185.799
PR-27341:  142.011

Pixel 3:
Before:    246.755
PR-27402:  160.045
PR-27341:  115.437

```

Single-threaded:

```
Pixel 2:
Before:    308.084
PR-27340:  303.539
PR-27341:  313.558

Pixel 3:
Before:    234.272
PR-27340:  227.158
PR-27341:  232.787

```

Differential Revision: [D17835333](https://our.internmc.facebook.com/intern/diff/D17835333)

Commit:
Modify PyTorch's integration of NNPACK to use a unified underlying thread pool implementation.

[ghstack-poisoned]

ljk53 commented on aten/src/ATen/native/NNPACK.cpp:
```diff
@@ -56,53 +56,39 @@ bool _nnpack_available() {
+}
+
+static pthreadpool_t nnpack_threadpool() {
+  return caffe2::mobile_pthreadpool();
```
mobile_pthreadpool() returns null for non-mobile build.
I don't know if NNPACK is used for non-mobile build at all - maybe @ezyang / @ilia-cher knows?
if we want to keep 100% the same behavior we should fallback to old thread creation logic for non-mobile build...

AshkanAliabadi replied:
Right, valid point.  I'll wait for Illia's input.

ezyang replied:
This test suggest nnpack is used in non-mobile:

```
    @unittest.skipIf(not torch._nnpack_available(), "NNPACK unavailable")
    def test_nnpack_conv(self):
        for kern, inp_size in [(3, 6), (3, 7), (4, 9)]:
            for batch, padding, chan_in, chan_out in \
                    product([1, 2], [0, 1, 2], [2], [3]):
```

ilia-cher replied:
let's fallback to a non-mobile thread pool

ljk53 commented on aten/src/ATen/native/NNPACK.cpp:
```diff
@@ -142,7 +128,7 @@ constexpr int weight_width_dim = 3;
 // constexpr int max_dim = 3;
 
-std::vector<int64_t> conv_output_size(
+static std::vector<int64_t> conv_output_size(
```
nit: I'd do this in a separate PR (as we might want to request pick to 1.3 we should keep this as small as possible).

AshkanAliabadi replied:
Good point, definitely.

ljk53 reviewed (commented):
Thanks for discovering and fixing this problem!

Commit:
Update on "Modify PyTorch's integration of NNPACK to use a unified underlying thread pool implementation."


Multi-threaded:

```
Pixel 2:
Before:    362.716
PR-27340:  185.799
PR-27341:  142.011

Pixel 3:
Before:    246.755
PR-27340:  160.045
PR-27341:  115.437

```

Single-threaded:

```
Pixel 2:
Before:    308.084
PR-27340:  303.539
PR-27341:  313.558

Pixel 3:
Before:    234.272
PR-27340:  227.158
PR-27341:  232.787

```

[ghstack-poisoned]

Commit:
Update on "Modify PyTorch's integration of NNPACK to use a unified underlying thread pool implementation."


Multi-threaded:

```
Pixel 2:
Before:    362.716
PR-27340:  185.799
PR-27341:  142.011

Pixel 3:
Before:    246.755
PR-27340:  160.045
PR-27341:  115.437

```

Single-threaded:

```
Pixel 2:
Before:    308.084
PR-27340:  303.539
PR-27341:  313.558

Pixel 3:
Before:    234.272
PR-27340:  227.158
PR-27341:  232.787

```

[ghstack-poisoned]

ljk53 commented on aten/src/ATen/native/NNPACK.cpp:
```diff
@@ -55,53 +55,67 @@ bool _nnpack_available() {
     nnpack_threadpool_ = pthreadpool_create(threads);
-    if (nnpack_threadpool_ == nullptr) {
-      throw std::runtime_error("could not initialize NNPack's pthreadpool");
+    if ( !nnpack_threadpool_ ) {
```
nit: remove extra space inside '()'

AshkanAliabadi replied:
Fixed it in 27402 to avoid having this PR rebuilt.

Commit:
Update on "Modify PyTorch's integration of NNPACK to use a unified underlying thread pool implementation."


Multi-threaded:

```
Pixel 2:
Before:    362.716
PR-27402:  185.799
PR-27341:  142.011

Pixel 3:
Before:    246.755
PR-27402:  160.045
PR-27341:  115.437

```

Single-threaded:

```
Pixel 2:
Before:    308.084
PR-27340:  303.539
PR-27341:  313.558

Pixel 3:
Before:    234.272
PR-27340:  227.158
PR-27341:  232.787

```

[ghstack-poisoned]

Commit:
Update on "Modify PyTorch's integration of NNPACK to use a unified underlying thread pool implementation."


Multi-threaded:

```
Pixel 2:
Before:    362.716
PR-27402:  185.799
PR-27341:  142.011

Pixel 3:
Before:    246.755
PR-27402:  160.045
PR-27341:  115.437

```

Single-threaded:

```
Pixel 2:
Before:    308.084
PR-27340:  303.539
PR-27341:  313.558

Pixel 3:
Before:    234.272
PR-27340:  227.158
PR-27341:  232.787

```

[ghstack-poisoned]
